src/thor/matrix_action.cc

```diff
@@ -81,32 +51,85 @@ std::string thor_worker_t::matrix(Api& request) {
     case TIME_DISTANCE_MATRIX:
-      matrix_algo = Matrix::TimeDistanceMatrix;
+      config_algo = Matrix::TimeDistanceMatrix;
       break;
   }
 
   // similar to routing: prefer the exact unidirectional algo if not requested otherwise
   // don't use matrix_type, we only need it to set the right warnings for what will be used
-  bool has_time =
-      check_matrix_time(request, options.prioritize_bidirectional() ? Matrix::CostMatrix
-                                                                    : Matrix::TimeDistanceMatrix);
-  if (has_time && !options.prioritize_bidirectional() && source_to_target_algorithm != COST_MATRIX) {
-    timedistancematrix();
-  } else if (has_time && options.prioritize_bidirectional() &&
+  if (has_time && !request.options().prioritize_bidirectional() &&
+      source_to_target_algorithm != COST_MATRIX) {
+    return &time_distance_matrix_;
+  } else if (has_time && request.options().prioritize_bidirectional() &&
              source_to_target_algorithm != TIME_DISTANCE_MATRIX) {
-    costmatrix(has_time);
-  } else if (matrix_algo == Matrix::CostMatrix) {
-    // if this happens, the server config only allows for timedist matrix
-    if (has_time && !options.prioritize_bidirectional()) {
+    return &costmatrix_;
+  } else if (config_algo == Matrix::CostMatrix) {
+    if (has_time && !request.options().prioritize_bidirectional()) {
       add_warning(request, 301);
     }
-    costmatrix(has_time);
+    return &costmatrix_;
   } else {
-    if (has_time && options.prioritize_bidirectional()) {
+    // if this happens, the server config only allows for timedist matrix
+    if (has_time && request.options().prioritize_bidirectional()) {
       add_warning(request, 300);
     }
-    timedistancematrix();
+    return &time_distance_matrix_;
   }
+}
+
+std::string thor_worker_t::matrix(Api& request) {
+  // time this whole method and save that statistic
+  auto _ = measure_scope_time(request);
+
+  auto& options = *request.mutable_options();
+  adjust_scores(options);
+  auto costing = parse_costing(request);
+
+  bool has_time =
+      check_matrix_time(request, options.prioritize_bidirectional() ? Matrix::CostMatrix
+                                                                    : Matrix::TimeDistanceMatrix);
+
+  // allow all algos to be cancelled
+  for (auto* alg : std::vector<MatrixAlgorithm*>{
+           &costmatrix_,
+           &time_distance_matrix_,
+           &time_distance_bss_matrix_,
+       }) {
+    alg->set_interrupt(interrupt);
+    alg->set_has_time(has_time);
+  }
+
+  auto* algo = get_matrix_algorithm(request, has_time, costing);
+  LOG_INFO("matrix::" + std::string(algo->name()));
+
+  // TODO(nils): TDMatrix doesn't care about either destonly or no_thru
+  if (algo->name() != "costmatrix") {
+    algo->SourceToTarget(request, *reader, mode_costing, mode,
+                         max_matrix_distance.find(costing)->second);
+    return tyr::serializeMatrix(request);
+  }
```

Comment:
we don't do second passes for tdmatrices (bss or normal), both don't have a concept of neither destonly nor not_thru

Reply:
actually unidir a* also allows for destonly, not sure why?